### PR TITLE
Support OCR provided patched WebLogic Server images

### PR DESCRIPTION
**Pull request by galiacheng**
This pr enables customer to deploy azure weblogic on aks offer with OCR provided patched WebLogic Server images.

### New experience
image

### Changes
- UI: support selecting image from middleware/weblogic_cpu repository
- Fail fast: validate OCR account and image with capability to check if the customer are granted to pull cpu images.
- Complete deployment script to support cpu images.
- Output: output base64 string of WebLogic Server versions
- Add pids to track deployments of different network option, LB/Application gateway/custom DNS

### Test case

- GA image
  - Pass: valid image tag: 14.1.1.0-11
  - Failed: invalid image tag: 14.1.1.0-13
- CPU image
  - Pass: have permission for cpu image + valid image tag from dropdown items: 12.2.1.4
  - Pass: have permission for cpu image + valid image tag provided by user: 12.2.1.4-slim-jdk8-ol7
  - Failed: OCR account has no permission for cpu images: use teams account
  - Failed: invalid image tag: 12.2.1.4-slim-jdk8-ol11
- User provided image: use image from ACR
- Outputs: base64 string of WLS version information

Workflow: https://github.com/galiacheng/weblogic-azure/actions/runs/1686017156

### CODE REVIEW

**Comment by edburns** [comment]
- Manually deploying to Partner Center test offer.
- Automated testing running at [Actions](https://github.com/edburns/weblogic-azure/actions/runs/1690185567)

**Comment by edburns** [comment]
@galiacheng  Please convert this to draft.

File: weblogic-azure-aks/src/main/arm/scripts/common.sh [hunk 1]
```diff
@@ -17,6 +17,9 @@ export constIntrospectorJobActiveDeadlineSeconds=300  # for Guaranteed Qos
 
 export curlMaxTime=120 # seconds
 export ocrLoginServer="container-registry.oracle.com"
+export ocrGaImagePath="middleware/weblogic"
+export ocrCpuImagePath="middleware/weblogic_cpu"
+export gitUrl4CpuImages="https://raw.githubusercontent.com/oracle/weblogic-azure/main/weblogic-azure-aks/src/main/resources/weblogic_cpu_images.json"
```
Comment (edburns) [inline weblogic-azure-aks/src/main/arm/scripts/common.sh:22]: This PR must be merged first: https://github.com/oracle/weblogic-azure/pull/105

**Comment by edburns** [comment]
This PR must be merged first: #105

File: weblogic-azure-aks/src/main/arm/scripts/inline-scripts/queryDomainConfigurations.sh [hunk 2]
```diff
@@ -23,41 +26,59 @@ domainConfigurationYaml=/tmp/domain.yaml
...
 if [ $? != 0 ]; then
-    echo >&2 "Fail to copy ${adminPodName}:/u01/wdt/models/model.properties."
+    echo >&2 "Fail to copy ${podName}:${targetFile4Versions}."
     exit 1
 fi
 
 base64ofDomainYaml=$(cat ${domainConfigurationYaml} | base64)
 base64ofModelYaml=$(cat ${targetModelYaml} | base64)
 base64ofModelProperties=$(cat ${targetModelProperties} | base64)
+base64ofWLSVersionDetails=$(cat ${targetFile4Versions} | base64)
```
Comment (edburns) [inline weblogic-azure-aks/src/main/arm/scripts/inline-scripts/queryDomainConfigurations.sh:75]: Yes!

File: weblogic-azure-aks/src/main/arm/scripts/inline-scripts/validateParameters.sh [hunk 3]
```diff
@@ -490,23 +551,23 @@ sslConfigurationAccessOption=$9
...
 enableAppGWIngress=${11}
 checkDNSZone=${12}
-checkACR=${13}
 
 ocrLoginServer="container-registry.oracle.com"
+ocrGaImagePath="middleware/weblogic"
+ocrCpuImagePath="middleware/weblogic_cpu"
 sslCertificateKeyVaultOption="keyVaultStoredConfig"
 userManagedIdentityType="Microsoft.ManagedIdentity/userAssignedIdentities"
+gitUrl4CpuImages="https://raw.githubusercontent.com/oracle/weblogic-azure/main/weblogic-azure-aks/src/main/resources/weblogic_cpu_images.json"
```
Comment (edburns) [inline weblogic-azure-aks/src/main/arm/scripts/inline-scripts/validateParameters.sh:560]: This appears to be a DRY violation. I see this same name=value in `common.sh`.  Can we source it from there?

**Review by edburns — CHANGES_REQUESTED** [review]
Just a couple small changes. But this PR should not be merged until #105 is merged anyway.  I want to avoid a situation where a PR references resources that don't already exist.

**Comment by galiacheng** [comment]
Resolve conflict

File: weblogic-azure-aks/src/main/arm/scripts/inline-scripts/validateParameters.sh (hunk 3 above)
Comment (galiacheng) [reply weblogic-azure-aks/src/main/arm/scripts/inline-scripts/validateParameters.sh]: Yes, Ed, I just found we can combine context of files in Deployment Script by `loadTextContent` of each file:

```
scriptContent: format('{0}\r\n\r\n{1}', loadTextContent('../../../arm/scripts/common.sh'), loadTextContent('../../../arm/scripts/inline-scripts/validateParameters.sh'))
```

see https://github.com/oracle/weblogic-azure/pull/104/commits/65a1d4b70cf9c09cf92b31d0d6d9f98164120009

**Comment by edburns** [comment]
Hello @mriccell I am seeking your review on the wording in the UI images pasted by @galiacheng  above, and also in this image, which is the tooltip for the `Is the specified SSO account associated with an active Oracle support contract?` question.

image

I'm not quite confident you would be comfortable with the use of "patched" in that text, but I could not come up with better text.

**Comment by mriccell** [comment]
The word patched/unpatched is strange in this context.  
I would say "... If 'Yes' is selected, the deployment process will pull from the CPU WebLogic Server image repository in the Oracle Container Registry (OCR). If 'No' is selected the deployment process will pull from the WebLogic Server image repository in OCR.

**Comment by galiacheng** [comment]
Hello @edburns  though we have agreed to add infobox to avoid empty Load Balancer configurations, after my investigation, we can filter the empty settings from UI output, which will not cause UI changes.

I achieved that using a long ARM function, see https://github.com/oracle/weblogic-azure/pull/104/commits/87e36951b07b32349fa1559dcf032c1db4374204

Please let me know if you prefer the experience we discussed in the standup meeting.

```
"lbSvcValues": "[if(not(empty(first(if(empty(steps('section_appGateway').lbSVCInfo.lbSVC),parse('[{\"colName\":\"\"}]'), steps('section_appGateway').lbSVCInfo.lbSVC)).colName)),steps('section_appGateway').lbSVCInfo.lbSVC, parse('[]'))]",

```

**Comment by galiacheng** [comment]
- Manually deploying to Partner Center test offer(v34).

**Comment by galiacheng** [comment]
Enhance UI for LB settings:
image

**Comment by galiacheng** [comment]
- Manually deploying to Partner Center test offer(v34).

**Review by mriccell — APPROVED** [review]

**Pull request merged**